---
title: "GSEA_v2"
author: "Natalia Garc√≠a"
date: '2023-05-16'
output: html_document
---
In this script we will

* Preprocess Gene Ontology (GO) gene set - "external gene name" (European Nucleoide Archive identifiers) annotations retrieved from BiomaRt API programmatic access - Ensembl Plants release 56 - _Brassica napus_ genome + labelled GO _Arabidopsis thaliana_  annotations from Reciprocal orthologous gene pairs Compara API in a set of selected categories of interest (See Annex I of Master thesis for more details ). Annotations are intended to be more comprehensive and specific than `PANTHERdb` or `PlantGSAD`, and Ensembl transcript centric contrary to those available in NCBI AnnotationHub R package.

* Conduct a FCS Gene Enrichment Analysis (GSEA) using the preranked GSEA algorithm with the package `genekitr`. 

# Loading libraries

```{r setup, include=FALSE}
suppressPackageStartupMessages(library("pathview"))
suppressPackageStartupMessages(library("readxl"))
suppressPackageStartupMessages(library("clusterProfiler"))
suppressPackageStartupMessages(library("Rgraphviz"))
suppressPackageStartupMessages(library("graph"))
library(tidyverse)
library(geneset)
library(genekitr)
library(data.table)
library("ggridges")
library("rrvgo")
library(gridExtra)
suppressPackageStartupMessages(library("KEGGgraph"))
suppressWarnings(suppressPackageStartupMessages(library("fgsea")))
```

Load available AnnotationHub Brassica napus annotation object to construct an Orgdb object for rrvgo to perform on

```{r}

library("AnnotationForge")
library("AnnotationDbi")
library("AnnotationHub")

# Making annotation package
ah <- AnnotationHub()

#Ensembl 56
query(ah, c( "OrgDb", "Brassica napus"))

org.Brassicanapus.eg.db <- ah[["AH107389"]]


```

# Loading data
logFC values
```{r}

logFC_HDAC<-read_excel("C:\\Users\\naata\\MASTER\\Enrichment\\Filtered_DEG_LFC1_padj05_results_total.xlsx")
logFC_Stress<-read_csv("C:\\Users\\naata\\MASTER\\Enrichment\\Final_condition_PErep_vs_VMrep_padj0.05_woPErep3.csv")
# merge external gene names
mappings<-read.csv("C:\\Users\\naata\\Downloads\\bna_ID_to_name.txt")
colnames(mappings)<-c("ensembl_gene_id","external_gene_name")
logFC_Stress<-merge(logFC_Stress,mappings, by="ensembl_gene_id")

# Import external gene name - log2C mappings
geneFC_HDAC<-logFC_HDAC[which(!duplicated(logFC_HDAC$external_gene_name)),3]
listfc_HDAC<-as.list(t(geneFC_HDAC))
names(listfc_HDAC) = as.list(t(logFC_HDAC[which(!duplicated(logFC_HDAC$external_gene_name)),8]))

geneFC_Stress<-logFC_Stress[which(!duplicated(logFC_Stress$external_gene_name)),3]
listfc_Stress<-as.list(t(geneFC_Stress))
names(listfc_Stress) = as.list(t(logFC_Stress[which(!duplicated(logFC_Stress$external_gene_name)),11]))

```

A collection of gene - DeSeq2 Wald statistic pair values is used in a `.rnk` file, sorted by the latter statistic representing gene expression levels

(Enter wald statistic formula : (/Zscore))

```{r cars}
ranks_HDAC <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_HDAC_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))

parents = c("GO:0040029","GO:0016570","GO:0034728","GO:0032776","GO:0080188","GO:0009850","GO:0009733","GO:0060918","GO:0008283","GO:0051301","GO:0016049","GO:0032502")
ranks_Stress <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_MVvsPE_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))


```
# LOAD genesets of interest from Excel file
```{r}
child_s_go <- read_excel("C:\\Users\\naata\\Documents\\MASTER\\TFM\\tfm\\GO_to_lookup_homologuesAt.xlsx", sheet="CHILD_TERMS_ALL")

parent_s_go <- read_excel("C:\\Users\\naata\\Documents\\MASTER\\TFM\\tfm\\GO_to_lookup_homologuesAt.xlsx", sheet="MAIN_TERMS")

colnames(child_s_go) <- c("ID", "name", "parent")
colnames(parent_s_go) <- c("ID", "name")


write.xlsx(BP_HDAC_list, "C://Users//naata//Desktop//reprex_plotGSEA.xlsx")
write.csv(gsea_Stress, "C://Users//naata//Desktop//GSEA_Stress_res.csv")
```

```{r}
lookup_terms_all = c("GO:0040029","GO:0016570","GO:0034728","GO:0032776","GO:0080188","GO:0016575","GO:0016574","GO:0016573","GO:0016571","GO:0031056","GO:0061647","GO:0016578","GO:0000412","GO:0004668","GO:0016577","GO:1990453","GO:0006334","GO:0006337","GO:0090124","GO:0090116","GO:0010424","GO:0010425","GO:0010426","GO:0141010","GO:0009850","GO:0009733","GO:0060918","GO:0090354","GO:0080024","GO:0090356","GO:0090355","GO:0010249","GO:0009683","GO:0009852","GO:0009851","GO:0071365","GO:0080026","GO:0009721","GO:0080162","GO:0010315","GO:0009926","GO:0060919","GO:0080161","GO:0008283","GO:0051301","GO:0016049","GO:0032502","GO:0042127","GO:0008284","GO:0008285","GO:0008356","GO:0051302","GO:0098725","GO:0051782","GO:0051781","GO:0009826","GO:0016475","GO:0048588","GO:0009825","GO:0030308","GO:0030307","GO:0001558","GO:0051210","GO:0051211","GO:0051093","GO:0051094","GO:0090644","GO:0030583","GO:0003006","GO:0098727","GO:0022611","GO:0009653","GO:0048532","GO:0060033","GO:0021700","GO:0010014","GO:0048856","GO:0031128","GO:0050793","GO:0009838","GO:0043696","GO:0043934","GO:0048589","GO:0048646","GO:0048869","GO:0090693","GO:0097737")

parents <- c("GO:0040029","GO:0016570","GO:0034728","GO:0032776","GO:0080188","GO:0009850","GO:0009733","GO:0060918","GO:0008283","GO:0051301","GO:0016049","GO:0032502")

ep_parents <- c("GO:0040029","GO:0016570","GO:0034728","GO:0032776")

lookup_epigenetic = c("GO:0040029","GO:0016570","GO:0034728","GO:0032776","GO:0080188","GO:0016575","GO:0016574","GO:0016573","GO:0016571","GO:0031056","GO:0061647","GO:0016578","GO:0000412","GO:0004668","GO:0016577","GO:1990453","GO:0006334","GO:0006337","GO:0090124","GO:0090116","GO:0010424","GO:0010425","GO:0010426","GO:0141010")

aux_parents <- c("GO:0009850","GO:0009733","GO:0060918")

lookup_auxin = c("GO:0090354","GO:0080024","GO:0090356","GO:0090355","GO:0010249","GO:0009683","GO:0009852","GO:0009851","GO:0071365","GO:0080026","GO:0009721","GO:0080162","GO:0010315","GO:0009926","GO:0060919","GO:0080161")

parents_emb <- c("GO:0008283","GO:0051301","GO:0016049","GO:0032502")

lookup_emb <- c("GO:0042127","GO:0008284","GO:0008285","GO:0008356","GO:0051302","GO:0098725","GO:0051782","GO:0051781","GO:0009826","GO:0016475","GO:0048588","GO:0009825","GO:0030308","GO:0030307","GO:0001558","GO:0051210","GO:0051211","GO:0051093","GO:0051094","GO:0090644","GO:0030583","GO:0003006","GO:0098727","GO:0022611","GO:0009653","GO:0048532","GO:0060033","GO:0021700","GO:0010014","GO:0048856","GO:0031128","GO:0050793","GO:0009838","GO:0043696","GO:0043934","GO:0048589","GO:0048646","GO:0048869","GO:0090693","GO:0097737")

lookup_development <-c("GO:0051093","GO:0051094","GO:0090644","GO:0030583","GO:0003006","GO:0098727","GO:0022611","GO:0009653","GO:0048532","GO:0060033","GO:0021700","GO:0010014","GO:0048856","GO:0031128","GO:0050793","GO:0009838","GO:0043696","GO:0043934","GO:0048589","GO:0048646","GO:0048869","GO:0090693","GO:0097737")

lookup_cell_division_prolif <- c("GO:0042127","GO:0008284","GO:0008285","GO:0008356","GO:0051302","GO:0098725","GO:0051782","GO:0051781")

lookup_cell_growth <- c("GO:0009826","GO:0016475","GO:0048588","GO:0009825","GO:0030308","GO:0030307","GO:0001558","GO:0051210","GO:0051211")

```

## Preprocessing expression table data into expression genelist format for fgsea

```{r pressure, echo=FALSE}


# Remove duplicates (transcript isoform with more differential expression will be conserved from the same gene)
ranks_HDAC <- ranks_HDAC[!duplicated(ranks_HDAC[,1]),]
ranks_Stress <- ranks_Stress[!duplicated(ranks_Stress[,1]),]

# Labelling columns (gene_name "ID" and statistic "t")
colnames(ranks_HDAC) <-c("ID","t")
colnames(ranks_Stress) <-c("ID","t")

# Sorting and creating ordered gene ranks column
ranks_HDAC <- ranks_HDAC %>% mutate(rank = rank(t,  ties.method = "random"))
ranks_HDAC <- ranks_HDAC[order(ranks_HDAC$rank, decreasing=TRUE), ]

ranks_Stress <- ranks_Stress %>% mutate(rank = rank(t,  ties.method = "random"))
ranks_Stress <- ranks_Stress[order(ranks_Stress$rank, decreasing=TRUE), ]

# Converting to set of lists
ranks_HDAC <- setNames(ranks_HDAC$t, ranks_HDAC$ID)
ranks_Stress <- setNames(ranks_Stress$t, ranks_Stress$ID)


```


# Loading pathways


# Loading gene set categories obtained in "get_gs_homolog_gs.Rmd"
```{r}

# Biological Process

pathways_BP_gs <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_bp_genesets.csv')
pathways_MF_gs <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_mf_genesets.csv')
pathways_CC_gs <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_cc_genesets.csv')

ml_BP_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_bp_genesets.rds")
ml_CC_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_cc_genesets.rds")
ml_MF_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_mf_genesets.rds")


# Handling pivoted multilevel lists and converting into gmt file format

# GO_BP
listindex <- seq(ml_BP_GO)
pathwayindex <- names(ml_BP_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_BP_GO[[i]])[[1]]
                        cbind(rep(names(ml_BP_GO)[i],length(gl)),gl)})

pathways_BP_gs<-do.call(rbind,  tmp_list)
colnames(pathways_BP_gs)=c("bp", "gene")

# Write to gmt
file.create("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_BP_formatted.gmt")
invisible(lapply(listindex, function(i){
  cat(names(ml_BP_GO)[i], "N/A", lapply(list(ml_BP_GO[[i]]), paste, collapse = "\t")[[1]], "\n", sep="\t", file="C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_BP_formatted.gmt", append=TRUE)}))




# GO_CC
listindex <- seq(ml_CC_GO)
pathwayindex <- names(ml_CC_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_CC_GO[[i]])[[1]]
                        cbind(rep(names(ml_CC_GO)[i],length(gl)),gl)})

# Write to gmt
pathways_CC_gs<-do.call(rbind,  tmp_list)
colnames(pathways_CC_gs)=c("cc", "gene")




file.create("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_CC_formatted.gmt")
invisible(lapply(listindex, function(i){
  cat(names(ml_CC_GO)[i], "N/A", lapply(list(ml_CC_GO[[i]]), paste, collapse = "\t")[[1]], "\n", sep="\t", file="C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_CC_formatted.gmt", append=TRUE)}))





# GO_MF
listindex <- seq(ml_MF_GO)
pathwayindex <- names(ml_MF_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_MF_GO[[i]])[[1]]
                        cbind(rep(names(ml_MF_GO)[i],length(gl)),gl)})

pathways_MF_gs<-do.call(rbind,  tmp_list)
colnames(pathways_MF_gs)=c("mf", "gene")

# Write to gmt
file.create("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_MF_formatted.gmt")
invisible(lapply(listindex, function(i){
  cat(names(ml_MF_GO)[i], "N/A", lapply(list(ml_MF_GO[[i]]), paste, collapse = "\t")[[1]], "\n", sep="\t", file="C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\Bna_GO_MF_formatted.gmt", append=TRUE)}))






```

Explotatory analysis of gene set distribution (informative to limit maximum gene set size in pathway enrichment)

```{r}
listindex <- seq(ml_BP_GO)
BP_gs_distribution <- lapply(listindex, function(i){length(list(ml_BP_GO[[i]])[[1]])})
BP_gs_distribution <- unlist(BP_gs_distribution)
BP_gs_distribution <- cbind(names(ml_BP_GO),BP_gs_distribution,rep("Biological Process", length(BP_gs_distribution)))

colnames(BP_gs_distribution)<- c("go", "GeneSetSize", "GeneOntology")



listindex <- seq(ml_MF_GO)
MF_gs_distribution <- lapply(listindex, function(i){length(list(ml_MF_GO[[i]])[[1]])})
MF_gs_distribution <- unlist(MF_gs_distribution)
MF_gs_distribution <- cbind(names(ml_MF_GO),MF_gs_distribution, rep("Molecular Function", length(MF_gs_distribution)))

colnames(MF_gs_distribution)<- c("go", "GeneSetSize", "GeneOntology")



listindex <- seq(ml_CC_GO)
CC_gs_distribution <- lapply(listindex, function(i){length(list(ml_CC_GO[[i]])[[1]])})
CC_gs_distribution <- unlist(CC_gs_distribution)
CC_gs_distribution <- cbind(names(ml_CC_GO),CC_gs_distribution, rep("Cellular Component", length(CC_gs_distribution)))

colnames(CC_gs_distribution)<-  c("go", "GeneSetSize", "GeneOntology")


library(reshape2)
df=as.data.frame(rbind(BP_gs_distribution,MF_gs_distribution, CC_gs_distribution)[,c(2,3)])
df$GeneOntology <- as.factor(df$GeneOntology)
df$GeneSetSize <- as.integer(df$GeneSetSize)


library("plyr")
ddply(df, "GeneOntology", summarise, grp.mean=mean(GeneSetSize), grp.std=sd(GeneSetSize), grp.median=median(GeneSetSize))

```
```{r}
df <- df[df$GeneSetSize > 2 ,]
df <- df[df$GeneSetSize < 2000 ,]
library(cowplot)
# Overlaid histograms
par(mfrow=c(2,1))
p1 <- ggplot(df, aes(x=GeneSetSize , color=GeneOntology)) +
  geom_histogram(color="black", fill="white", bins = 100)+
  facet_grid(~GeneOntology)

p2 <- ggplot(df, aes(x=GeneSetSize)) +
  geom_histogram(color="red", fill="white", bins = 100)
plot_grid(p1 , p2, ncol=1, align="v")
```


Loading pathway term descriptions and selected GO categories of interest (MAIN ANCESTOR and their respective CHILD terms)

```{r}

# Loading term descriptions
pathways_BP_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_bp_terms.csv')
pathways_MF_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_mf_terms.csv')
pathways_CC_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_cc_terms.csv')

# Loading selected categories
child_s_go <- read_excel("C:\\Users\\naata\\Documents\\MASTER\\TFM\\tfm\\GO_to_lookup_homologuesAt.xlsx", sheet="CHILD_TERMS_ALL")
parent_s_go <- read_excel("C:\\Users\\naata\\Documents\\MASTER\\TFM\\tfm\\GO_to_lookup_homologuesAt.xlsx", sheet="MAIN_TERMS")

colnames(child_s_go) <- c("ID", "name", "parent")
colnames(parent_s_go) <- c("ID", "name")


```

# Gene set enrichment analysis

## Biological process GSEA

```{r}
library("clusterProfiler")
library(stringr)

attr<- c("ID","Description","setSize","enrichmentScore","NES","pvalue","p.adjust","qvalue","rank","leading_edge","geneID", "Count","GeneRatio")

gsea_HDAC_o <- clusterProfiler::GSEA(geneList = ranks_HDAC,nPerm =10000, minGSSize=5,maxGSSize=2000, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = pathways_BP_gs, TERM2NAME = pathways_BP_terms, seed=TRUE, by="fgsea")
gsea_HDAC<-gsea_HDAC_o[,1:11]

gsea_Stress_o <- clusterProfiler::GSEA(geneList = ranks_Stress,nPerm =10000, minGSSize=5,maxGSSize=2000, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = pathways_BP_gs, TERM2NAME = pathways_BP_terms, seed=TRUE, by="fgsea")
gsea_Stress<-gsea_Stress_o[,1:11]



```
### Saving result 
```{r}
write.csv(gsea_HDAC, "C://Users//naata//Documents//MASTER//TFM//GSEA_HDACi_res_all_gs.csv")
write.csv(gsea_Stress, "C://Users//naata//Documents//MASTER//TFM//GSEA_Stress_res_all_gs.csv")
```

### Saving overrepresented terms

```{r}
write.table(gsea_HDAC[which(gsea_HDAC$p.adjust<0.05),]$ID, "C://Users//naata//Documents//MASTER//TFM//GSEA_HDACi_res_BP_terms.txt")
write.table(gsea_Stress[which(gsea_Stress$p.adjust<0.05),]$ID, "C://Users//naata//Documents//MASTER//TFM//GSEA_Stress_res_BP_terms.txt")

```



# BP visualizations



Preprocessing genekitr for plotting

```{r}

attr<-c("ID", "Description", "setSize", "enrichmentScore","NES","pvalue","p.adjust","qvalue",  "rank","leading_edge","geneID","geneID_symbol","Count","GeneRatio" )
gsea_HDAC = gsea_HDAC_o[,1:11]
gsea_Stress = gsea_Stress_o[,1:11]

count_HDAC <- c()
gsea_HDAC$geneID_symbol = gsea_HDAC$core_enrichment
for (x in gsea_HDAC$core_enrichment){count_HDAC<-c(count_HDAC, as.numeric(str_count(x, "/")+1))}
gsea_HDAC$Count <- count_HDAC
gsea_HDAC$GeneRatio <- gsea_HDAC$Count/gsea_HDAC$setSize
names(gsea_HDAC) <- attr

count_Stress <- c()
gsea_Stress$geneID_symbol = gsea_Stress$core_enrichment
for (x in gsea_Stress$core_enrichment){count_Stress<-c(count_Stress, as.numeric(str_count(x, "/")+1))}
gsea_Stress$Count <- count_Stress
gsea_Stress$GeneRatio <- gsea_Stress$Count/gsea_Stress$setSize
names(gsea_Stress) <- attr

# Filter by padj
gsea_HDAC_f = gsea_HDAC[which(gsea_HDAC$p.adjust<0.05),]
gsea_Stress_f = gsea_Stress[which(gsea_Stress$p.adjust<0.05),]


# LFC

lfc_val_HDAC <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_HDAC_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))
lfc_val_Stress <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_MVvsPE_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))

lfc_val_HDAC <- lfc_val_HDAC[!duplicated(lfc_val_HDAC[,1]),]
lfc_val_Stress <- lfc_val_Stress[!duplicated(lfc_val_Stress[,1]),]
colnames(lfc_val_HDAC) <- c("ID", "logfc")
colnames(lfc_val_Stress) <- c("ID", "logfc")

# METADATA

org_df <- data.frame(org="bnapus")
exp_df <- data.frame(exponent=1)

# GSEA result

pathways_BP_gs <- as.data.frame(pathways_BP_gs)
colnames(pathways_BP_gs) <- c("bp","name")

BP_HDAC_list = list()
BP_HDAC_list[['gsea_df']] <- gsea_HDAC_f
BP_HDAC_list[['genelist']] <- lfc_val_HDAC
BP_HDAC_list[['geneset']] <- pathways_BP_gs
#BP_HDAC_list[['geneset']] <- pathways_BP_gs[!(pathways_BP_gs[,1] %in% c(child_s_go[,1],parent_s_go[,1])),]
BP_HDAC_list[['exponent']] <- exp_df
BP_HDAC_list[['org']] <- org_df

BP_Stress_list = list()
BP_Stress_list[['gsea_df']] <- gsea_Stress_f
BP_Stress_list[['genelist']] <- lfc_val_Stress
BP_Stress_list[['geneset']] <- pathways_BP_gs
#BP_Stress_list[['geneset']] <- pathways_BP_gs[!(pathways_BP_gs[,1] %in% c(child_s_go[,1],parent_s_go[,1])),]
BP_Stress_list[['exponent']] <- exp_df
BP_Stress_list[['org']] <- org_df

```

# Semantic analysis of GO terms involved in BP with rrvgo
```{r}
##### HDACi
# Calculate similarity matrix
simMatrix_HDACi_gsea <- calculateSimMatrix(gsea_HDAC_f$ID,
                                orgdb=org.Brassicanapus.eg.db,
                                ont="BP",
                                method="Wang")
# Reducing terms
scores <- setNames(-log10(gsea_HDAC_f$qvalue), gsea_HDAC_f$ID)
reducedTerms_HDAC_gsea <- reduceSimMatrix(simMatrix_HDACi_gsea,
                                scores,
                                threshold=0.7,
                                orgdb=org.Brassicanapus.eg.db)


##### STRESS

# Calculate similarity matrix
simMatrix_Stress_gsea <- calculateSimMatrix(gsea_Stress_f$ID,
                                orgdb=org.Brassicanapus.eg.db,
                                ont="BP",
                                method="Wang")
# Reducing terms
scores <- setNames(-log10(gsea_Stress_f$qvalue), gsea_Stress_f$ID)
reducedTerms_Stress_gsea <- reduceSimMatrix(simMatrix_Stress_gsea,
                                scores,
                                threshold=0.7,
                                orgdb=org.Brassicanapus.eg.db)
```




```{r}
gsea_HDAC_f_up <- gsea_HDAC_f[gsea_HDAC_f$enrichmentScore >0,]
gsea_HDAC_f_down <- gsea_HDAC_f[gsea_HDAC_f$enrichmentScore <0,]

gsea_Stress_f_up <- gsea_Stress_f[gsea_Stress_f$enrichmentScore >0,]
gsea_Stress_f_down <- gsea_Stress_f[gsea_Stress_f$enrichmentScore <0,]

# getting parent pathways from GO-hierarchy-supervised semantic clustering.
gsea_Stress_up_parent <- gsea_Stress_f_up[gsea_Stress_f_up$ID %in% unique(reducedTerms_Stress_gsea[,3]),]
gsea_Stress_down_parent <- gsea_Stress_f_down[gsea_Stress_f_down$ID %in% unique(reducedTerms_Stress_gsea[,3]),]

gsea_HDAC_up_parent <- gsea_HDAC_f_up[gsea_HDAC_f_up$ID %in% unique(reducedTerms_HDAC_gsea[,3]),]
gsea_HDAC_down_parent <- gsea_HDAC_f_down[gsea_HDAC_f_down$ID %in% unique(reducedTerms_HDAC_gsea[,3]),]

```
Save leading edge genes

```{r}
# Rbind parent categories and lookup categories
Stress_filtered_lookup_up<-rbind(gsea_Stress_up_parent, gsea_Stress_f_up[gsea_Stress_f_up$ID %in% lookup_terms_all,])
Stress_filtered_lookup_down<-rbind(gsea_Stress_down_parent, gsea_Stress_f_down[gsea_Stress_f_down$ID %in% lookup_terms_all,])


HDAC_filtered_lookup_up<-rbind(gsea_HDAC_up_parent, gsea_HDAC_f_up[gsea_HDAC_f_up$ID %in% lookup_terms_all,])
HDAC_filtered_lookup_down<-rbind(gsea_HDAC_down_parent, gsea_HDAC_f_down[gsea_HDAC_f_down$ID %in% lookup_terms_all,])

# Get leading edge genes from parents and lookup categories
unlist_le_genes <- function(x) {unlist(strsplit(x, "/"))}

le_all<-c()

for (i in Stress_filtered_lookup_up$geneID){
  newlist <- unlist_le_genes(i)
  le_all <- c(le_all, newlist)
  le_all <- unique(le_all)
}

write.table(le_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\le_Stress_up.txt")


le_all<-c()
for (i in Stress_filtered_lookup_down$geneID){
  newlist <- unlist_le_genes(i)
  le_all <- c(le_all, newlist)
  le_all <- unique(le_all)
}

write.table(le_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\le_Stress_down.txt")


le_all<-c()
for (i in HDAC_filtered_lookup_up$geneID){
  newlist <- unlist_le_genes(i)
  le_all <- c(le_all, newlist)
  le_all <- unique(le_all)
}

write.table(le_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\le_HDAC_up.txt")

le_all<-c()
for (i in HDAC_filtered_lookup_down$geneID){
  newlist <- unlist_le_genes(i)
  le_all <- c(le_all, newlist)
  le_all <- unique(le_all)
}

write.table(le_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\le_HDAC_down.txt")

```





# Classic plot


# ALL parent terms

```{r}

p1 <- genekitr::plotGSEA(BP_HDAC_list, plot_type = "classic", show_pathway = which(BP_HDAC_list$gsea_df$ID %in% parents), border_thick = 1.3, main_text_size = 9, legend_text_size = 8, wrap_length = TRUE, label_by = 'description')
p2 <- genekitr::plotGSEA(BP_Stress_list, plot_type = "classic", show_pathway = which(BP_Stress_list$gsea_df$ID %in% parents), border_thick = 1.3, main_text_size = 9, legend_text_size = 8, wrap_length = TRUE, label_by = 'description')

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\classic_parent_comparison.png", units="in", width=13, height=10, res=300)
grid.arrange(p1 , p2)
dev.off()
```

Epigenetic terms 

```{r}
# red palette
p1 <- genekitr::plotGSEA(BP_HDAC_list, plot_type = "classic", show_pathway = which(BP_HDAC_list$gsea_df$ID %in% lookup_epigenetic), border_thick = 1.3, main_text_size = 9, legend_text_size = 8, wrap_length = TRUE)
p2 <- genekitr::plotGSEA(BP_Stress_list, plot_type = "classic", show_pathway = which(BP_Stress_list$gsea_df$ID %in% lookup_epigenetic), border_thick = 1.3, main_text_size = 9, legend_text_size = 8)

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\classic_epig_comparison.png", units="in", width=7, height=10, res=300)
grid.arrange(p1 , p2)
dev.off()
```

# Developmental pathways
```{r}
# orange palette
p1 <- genekitr::plotGSEA(BP_HDAC_list, plot_type = "classic", show_pathway = which(BP_HDAC_list$gsea_df$ID %in% lookup_emb), border_thick = 1.3, main_text_size = 9, legend_text_size = 8, wrap_length = TRUE)
p2 <- genekitr::plotGSEA(BP_Stress_list, plot_type = "classic", show_pathway = which(BP_Stress_list$gsea_df$ID %in% lookup_emb), border_thick = 1.3, main_text_size = 9, legend_text_size = 8)

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\classic_dev_comparison.png", units="in", width=7, height=10, res=300)
grid.arrange(p1 , p2)
dev.off()
```

# Auxin (only overrepresented in stress)

```{r}

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\classic_aux_comparison.png", units="in", width=7, height=6, res=300)
genekitr::plotGSEA(BP_Stress_list, plot_type = "classic", show_pathway = which(BP_Stress_list$gsea_df$ID %in% lookup_auxin), border_thick = 1.3, main_text_size = 9, legend_text_size = 8)
dev.off()
```

# cell growth (only overrepresented in stress)
```{r}

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\classic_cg_comparison.png", units="in", width=7, height=6, res=300)
genekitr::plotGSEA(BP_Stress_list, plot_type = "classic", show_pathway = which(BP_Stress_list$gsea_df$ID %in% lookup_cell_growth), border_thick = 1.3, main_text_size = 9, legend_text_size = 8)
dev.off()
```

```{r}
gsea_Stress_f[gsea_Stress_f$ID %in% lookup_terms_all,]
gsea_HDAC_f[gsea_HDAC_f$ID %in% lookup_terms_all,]
```



```{r}
pathways = as.data.frame(gsea_HDAC[which(gsea_HDAC[,1] %in% parent_s_go[,1]$ID),1])
library(patchwork)
p1 <- plotGSEA(BP_HDAC_list, plot_type = "fgsea", show_pathway = pathways[1:12,1])
p2 <- plotGSEA(BP_Stress_list, plot_type = "fgsea", show_pathway = pathways[1:12,1])
p1  + plot_annotation(tag_levels = "A")

```

```{r}

# Top 15 terms ridge plot
p1<- plotGSEA(BP_HDAC_list,
  plot_type = "ridge",
  show_pathway = which(BP_HDAC_list$gsea_df$ID %in% lookup_terms_all), stats_metric = "p.adjust", border_thick = 1.3, main_text_size = 12, legend_text_size = 8
) + labs(title = "HDACi PE selected terms", tag ="A")

p2<-plotGSEA(BP_Stress_list,
  plot_type = "ridge",
  show_pathway = which(BP_Stress_list$gsea_df$ID %in% lookup_terms_all), stats_metric = "p.adjust", border_thick = 1.3, main_text_size = 12, legend_text_size = 8
) + labs(title = "ME selected terms", tag ="B")

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\ridge_lookup_comparison.png", units="in", width=5, height=10, res=300)
grid.arrange(p1 , p2)
dev.off()
```
```{r}
## two-side barplot


p1<- plotGSEA(BP_HDAC_list, plot_type = "bar",  show_pathway = which(BP_HDAC_list$gsea_df$ID %in% parents), stats_metric = "p.adjust", border_thick = 1.3, main_text_size = 12, legend_text_size = 8, colour = c("green", "red")) + labs(title = "HDACi PE selected terms", tag ="A")

p2<-plotGSEA(BP_Stress_list, plot_type = "bar",
  show_pathway = which(BP_Stress_list$gsea_df$ID %in% parents), stats_metric = "p.adjust", border_thick = 1.3, main_text_size = 12, legend_text_size = 8, colour = c("green", "red")
) + labs(title = "ME selected terms", tag ="B")

png("C:\\Users\\naata\\Documents\\MASTER\\TFM\\bar_parent_comparison.png", units="in", width=5, height=10, res=300)
grid.arrange(p1 , p2)
dev.off()

```



















# Filter by pval and fdr q-value, plot all resulting terms


```{r}
x <- filter(gsea_HDAC_o, p.adjust < .05, qvalue < 0.25)

x_up <- filter(x, enrichmentScore>0)
x_down <- filter(x, enrichmentScore<0)
x_df_up <- x_up[,1:11]
x_df_down <- x_down[,1:11]

pathways_up = as.data.frame(x_df_up[which(x_df_up[,1] %in% c(parent_s_go[,1]$ID, child_s_go[,1]$ID)),1])

pathways_down = as.data.frame(x_df_down[which(x_df_down[,1] %in% c(parent_s_go[,1]$ID, child_s_go[,1]$ID)),1])

```

```{r}
# Plot up-reg gs of interest
gseaplot2(x_up, geneSetID = which(x_up[,1] %in% (pathways_up[,1])), pvalue_table=TRUE, subplots= 1:3)
```

```{r}
# Plot down-reg gs of interest
gseaplot2(x_down, geneSetID = which(x_down[,1] %in% (pathways_down[,1])), pvalue_table=TRUE, subplots= 1:3)
```



```{r}
x <- filter(gsea_Stress_o, p.adjust < .05, qvalue < 0.25)
x_df <- x[,1:11]


pathways = as.data.frame(x_df[which(x_df[,1] %in% c(parent_s_go[,1]$ID, child_s_go[,1]$ID)),1])


gseaplot2(x, geneSetID = which(x[,1] %in% (pathways[,1])), pvalue_table=TRUE, subplots= 1:3)
```

```{r}

```




































