---
title: "ORA_analysis"
author: "Natalia Garc√≠a"
date: '2023-05-17'
output: html_document
---

Overrepresentation analysis of genesets


* Conduct an overrepresentation analysis hyper-geometric - exact Fisher test with FDR correction for Gene Enrichment Analysis using clusterProfiler

Collapse pathways
Save results
Visualize lookup genes

conduct for PLANTGSAD (TF, Cyc, KEGG)

# Loading libraries

```{r setup, include=FALSE}
suppressPackageStartupMessages(library("pathview"))
suppressPackageStartupMessages(library("readxl"))
suppressPackageStartupMessages(library("clusterProfiler"))
suppressPackageStartupMessages(library("Rgraphviz"))
suppressPackageStartupMessages(library("graph"))
library(tidyverse)
library(geneset)
library(genekitr)
library(data.table)
suppressPackageStartupMessages(library("KEGGgraph"))
suppressWarnings(suppressPackageStartupMessages(library("fgsea")))
```

Load available AnnotationHub Brassica napus annotation object to construct an Orgdb object for rrvgo to perform on

```{r}

library("AnnotationForge")
library("AnnotationDbi")
library("AnnotationHub")
library("OrgDB")

# Making annotation package
ah <- AnnotationHub()

#Ensembl 56
query(ah, c( "OrgDb", "Brassica napus"))

org.Brassicanapus.eg.db <- ah[["AH107389"]]


```

# Loading data

A collection of gene - DeSeq2 Wald statistic pair values is used in a `.rnk` file, sorted by the latter statistic representing gene expression levels

(Enter wald statistic formula : (/Zscore))

```{r cars}
ranks_HDAC <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_HDAC_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))


ranks_Stress <- read.csv("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\bna_MVvsPE_deseq2stat.rnk",
                    header=FALSE, colClasses = c("character", "numeric"))

# We'll extract the background genes - the genes detected with over log10(counts)>1

background_ALL_gl_HDAC <- as.data.frame(ranks_HDAC[,1])
background_ALL_gl_Stress <- as.data.frame(ranks_Stress[,1])

background_UP_gl_HDAC <- as.data.frame(ranks_HDAC[which(ranks_HDAC[,2] > 0), 1])
background_UP_gl_Stress <- as.data.frame(ranks_Stress[which(ranks_Stress[,2] > 0), 1])

background_DOWN_gl_HDAC <- as.data.frame(ranks_HDAC[which(ranks_HDAC[,2] < 0), 1])
background_DOWN_gl_Stress <- as.data.frame(ranks_Stress[which(ranks_Stress[,2] < 0), 1])

```

## Preprocessing data

```{r pressure, echo=FALSE}


# Remove duplicates (transcript isoform with more differential expression will be conserved from the same gene)
ranks_HDAC <- ranks_HDAC[!duplicated(ranks_HDAC[,1]),]
ranks_Stress <- ranks_Stress[!duplicated(ranks_Stress[,1]),]

# Labelling columns (gene_name "ID" and statistic "t")
colnames(ranks_HDAC) <-c("ID","t")
colnames(ranks_Stress) <-c("ID","t")

# Sorting and creating ordered gene ranks column
ranks_HDAC <- ranks_HDAC %>% mutate(rank = rank(t,  ties.method = "random"))
ranks_HDAC <- ranks_HDAC[order(ranks_HDAC$rank, decreasing=TRUE), ]

ranks_Stress <- ranks_Stress %>% mutate(rank = rank(t,  ties.method = "random"))
ranks_Stress <- ranks_Stress[order(ranks_Stress$rank, decreasing=TRUE), ]

# Converting to set of lists
ranks_HDAC <- setNames(ranks_HDAC$t, ranks_HDAC$ID)
ranks_Stress <- setNames(ranks_Stress$t, ranks_Stress$ID)


```

# Loading pathways and background gene set


# Loading gene set categories obtained in "get_gs_homolog_gs.Rmd"
```{r}

# Biological Process

ml_BP_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_bp_genesets.rds")
ml_CC_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_cc_genesets.rds")
ml_MF_GO <- readRDS("C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_mf_genesets.rds")

# Handling pivoted multilevel lists and converting into gmt file format

# GO_BP
listindex <- seq(ml_BP_GO)
pathwayindex <- names(ml_BP_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_BP_GO[[i]])[[1]]
                        cbind(rep(names(ml_BP_GO)[i],length(gl)),gl)})

pathways_BP_gs<-do.call(rbind,  tmp_list)
colnames(pathways_BP_gs)=c("bp", "gene")


# GO_CC
listindex <- seq(ml_CC_GO)
pathwayindex <- names(ml_CC_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_CC_GO[[i]])[[1]]
                        cbind(rep(names(ml_CC_GO)[i],length(gl)),gl)})

# Write to gmt
pathways_CC_gs<-do.call(rbind,  tmp_list)
colnames(pathways_CC_gs)=c("cc", "gene")


# GO_MF
listindex <- seq(ml_MF_GO)
pathwayindex <- names(ml_MF_GO)

# Pass to df format
tmp_list <- lapply(listindex, function(i){
                        gl = list(ml_MF_GO[[i]])[[1]]
                        cbind(rep(names(ml_MF_GO)[i],length(gl)),gl)})

pathways_MF_gs<-do.call(rbind,  tmp_list)
colnames(pathways_MF_gs)=c("mf", "gene")


pathways_BP_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_bp_terms.csv')
pathways_MF_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_mf_terms.csv')
pathways_CC_terms <-read.csv('C:\\Users\\naata\\Documents\\MASTER\\BioSynth\\synthetic_env\\GSEA\\full_cc_terms.csv')

```

# OVERREPRESENTATION ANALYSIS

# Import results from DeSeq2 analysis
```{r}
DEG_HDAC<-read_excel("C:\\Users\\naata\\MASTER\\Enrichment\\Filtered_DEG_LFC1_padj05_results_total.xlsx")
DEG_Stress <-read_csv("C:\\Users\\naata\\MASTER\\Enrichment\\Final_condition_PErep_vs_VMrep_padj0.05_woPErep3_LFC1filtered.csv")

DEG_ALL_HDAC <- DEG_HDAC$external_gene_name
DEG_ALL_Stress <- DEG_Stress$external_gene_name

DEG_UP_HDAC <- DEG_HDAC[which(DEG_HDAC$log2FoldChange > 0),]$external_gene_name
DEG_UP_Stress <- DEG_Stress[which(DEG_Stress$log2FoldChange > 0),]$external_gene_name


DEG_DOWN_HDAC <- DEG_HDAC[which(DEG_HDAC$log2FoldChange < 0),]$external_gene_name
DEG_DOWN_Stress <- DEG_Stress[which(DEG_Stress$log2FoldChange < 0),]$external_gene_name

```

# All genes
```{r}
colnames(pathways_BP_gs)<-c("bp", "gene")

# Get background genes : genes with detected minimal expression
colnames(background_ALL_gl_HDAC) = "gene"
bckgr_geneset_HDAC <- merge(background_ALL_gl_HDAC, pathways_BP_gs, by="gene")

colnames(background_ALL_gl_Stress) = "gene"
bckgr_geneset_Stress <- merge(background_ALL_gl_Stress, pathways_BP_gs, by="gene")

colnames(bckgr_geneset_HDAC)<-c("ID","bp")
bckgr_geneset_HDAC<-bckgr_geneset_HDAC[,c(2,1)]
colnames(bckgr_geneset_Stress)<-c("ID","bp")
bckgr_geneset_Stress<-bckgr_geneset_Stress[,c(2,1)]

ORA_HDAC_ALL <- clusterProfiler::enricher(gene = DEG_ALL_HDAC, minGSSize=2, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_HDAC, TERM2NAME = pathways_BP_terms)
ORA_Stress <- clusterProfiler::enricher(gene = DEG_ALL_Stress, minGSSize=2, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_Stress, TERM2NAME = pathways_BP_terms)

ORA_HDAC_ALL@organism="bnapus"
ORA_HDAC_ALL@ontology="BP"

ORA_Stress@organism="bnapus"
ORA_Stress@ontology="BP"





HDAC_all <- importCP(ORA_HDAC_ALL, type = "other")
write_csv(HDAC_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\ORA_HDAC_ALL_Results.csv")

Stress_all <- importCP(ORA_Stress, type = "other")
write_csv(Stress_all, "C:\\Users\\naata\\Documents\\MASTER\\TFM\\ORA_Stress_ALL_Results.csv")
```

Post processing output and calculating FoldEnrich and RichFactor output
```{r}

ORA_HDAC_ALL = ORA_HDAC_ALL[,]
ORA_Stress = ORA_Stress[,]

calcratios<-function(x){vec_string<-as.integer(unlist(strsplit( x, split = "/")))
                        vec_string[1] / vec_string[2]}

##### HDAC

generat_vector<-unlist(lapply(ORA_HDAC_ALL$GeneRatio,calcratios))
bg_vector<-unlist(lapply(ORA_HDAC_ALL$BgRatio,calcratios))
gcount_in_bg <- unlist(lapply(ORA_HDAC_ALL$BgRatio, function(x){vec_string<-as.integer(unlist(strsplit( x, split = "/")))
                        vec_string[1]}))

# Calculate ratios
ORA_HDAC_ALL$GeneRatio<-generat_vector
ORA_HDAC_ALL$BgRatio<-bg_vector

# Add Fold Enrichment = GeneRatio /  BgRatio
ORA_HDAC_ALL$FoldEnrich<-generat_vector/bg_vector
  
# Add rich factor = Gcount in set / Gcount in Bg
ORA_HDAC_ALL$RichFactor<-ORA_HDAC_ALL$Count/gcount_in_bg


##### STRESS
generat_vector<-unlist(lapply(ORA_Stress$GeneRatio,calcratios))
bg_vector<-unlist(lapply(ORA_Stress$BgRatio,calcratios))
gcount_in_bg <- unlist(lapply(ORA_Stress$BgRatio, function(x){vec_string<-as.integer(unlist(strsplit( x, split = "/")))
                        vec_string[1]}))
# Calculate ratios
ORA_Stress$GeneRatio<-generat_vector
ORA_Stress$BgRatio<-bg_vector

# Add Fold Enrichment = GeneRatio /  BgRatio
ORA_Stress$FoldEnrich<-generat_vector/bg_vector

# Add rich factor = Gcount in set / Gcount in Bg
ORA_Stress$RichFactor<-ORA_Stress$Count/gcount_in_bg


write_csv(ORA_Stress[,], "C:\\Users\\naata\\Documents\\MASTER\\TFM\\ORA_Stress_ALL_Results_CP.csv")
write_csv(ORA_HDAC_ALL[,], "C:\\Users\\naata\\Documents\\MASTER\\TFM\\ORA_HDAC_ALL_Results_CP.csv")

```


Rrvigo Wang method semantic similarity calculation and plotting/clustering terms


```{r}
library("rrvgo")
HDAC_sign<-ORA_HDAC_ALL[ORA_HDAC_ALL$p.adjust<0.05,]
Stress_sign<-ORA_Stress[ORA_Stress$p.adjust<0.05,]

##### HDACi
# Calculate similarity matrix
simMatrix_HDACi <- calculateSimMatrix(HDAC_sign$ID,
                                orgdb=org.Brassicanapus.eg.db,
                                ont="BP",
                                method="Wang")
# Reducing terms
scores <- setNames(-log10(HDAC_sign$qvalue), HDAC_sign$ID)
reducedTerms_HDAC <- reduceSimMatrix(simMatrix_HDACi,
                                scores,
                                threshold=0.7,
                                orgdb=org.Brassicanapus.eg.db)


##### STRESS

# Calculate similarity matrix
simMatrix_Stress <- calculateSimMatrix(Stress_sign$ID,
                                orgdb=org.Brassicanapus.eg.db,
                                ont="BP",
                                method="Wang")
# Reducing terms
scores <- setNames(-log10(Stress_sign$qvalue), Stress_sign$ID)
reducedTerms_Stress <- reduceSimMatrix(simMatrix_Stress,
                                scores,
                                threshold=0.7,
                                orgdb=org.Brassicanapus.eg.db)
```
# Similarity matrix

```{r}

p1 <- heatmapPlot(simMatrix_HDACi,
            reducedTerms_HDAC,
            annotateParent=TRUE,
            annotationLabel="parentTerm",
            fontsize=6)
p2 <- heatmapPlot(simMatrix_Stress,
            reducedTerms_Stress,
            annotateParent=TRUE,
            annotationLabel="parentTerm",
            fontsize=6)
library(gridExtra)
png(filename="C:\\Users\\naata\\Documents\\MASTER\\TFM\\cluster_ORA_HDAC.png", width=25,height=20, units="cm" ,res=300)
p1
dev.off()

```
```{r}
p1<- treemapPlot(reducedTerms_HDAC)
p2<- treemapPlot(reducedTerms_Stress)
p2
```


```{r}

# scatterplot
p1<- scatterPlot(simMatrix_HDACi, reducedTerms_HDAC, size="size", addLabel = TRUE)
p2<- scatterPlot(simMatrix_Stress, reducedTerms_Stress, size="size", addLabel = TRUE)

png(filename="C:\\Users\\naata\\Documents\\MASTER\\TFM\\scatterplot_ORA_HDAC.png", width=25,height=20, units="cm" ,res=300)
grid
dev.off()

```
```{r}

p1<- wordcloudPlot(reducedTerms_HDAC, min.freq=1, colors="black")
p2<- scatterPlot(simMatrix_Stress, reducedTerms_Stress, size="size", addLabel = TRUE)


# wordcloud


```


```{r}

genekitr::plotEnrich(go_easy, fold_change = listfc_HDAC, plot_type = "goheat", sim_method = "Wang", org = "bnapus", ont="BP")

```



# Up-regulated genes

```{r}
library("clusterProfiler")
colnames(background_UP_gl_HDAC) = "gene"
bckgr_geneset_HDAC <- merge(background_UP_gl_HDAC, pathways_BP_gs, by="gene")

colnames(background_ALL_gl_Stress) = "gene"
bckgr_geneset_Stress <- merge(background_UP_gl_Stress, pathways_BP_gs, by="gene")

ORA_HDAC_UP <- clusterProfiler::enricher(geneList = DEG_UP_HDAC, minGSSize=2, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_HDAC, TERM2NAME = ml_BP_GO, seed=TRUE, by="fgsea")
ORA_Stress_UP <- clusterProfiler::enricher(geneList = DEG_UP_Stress, minGSSize=2, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_Stress, TERM2NAME = ml_BP_GO, seed=TRUE, by="fgsea")


```

# Down-regulated genes

```{r}
library("clusterProfiler")
colnames(background_ALL_gl_HDAC) = "gene"
bckgr_geneset_HDAC <- merge(background_DOWN_gl_HDAC, pathways_BP_gs, by="gene")

colnames(background_ALL_gl_Stress) = "gene"
bckgr_geneset_Stress <- merge(background_DOWN_gl_Stress, pathways_BP_gs, by="gene")

ORA_HDAC_ALL <- clusterProfiler::enricher(geneList = DEG_DOWN_HDAC, minGSSize=2, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_HDAC, TERM2NAME = pathways_BP_terms, seed=TRUE, by="fgsea")
ORA_Stress <- clusterProfiler::enricher(geneList = DEG_DOWN_Stress, minGSSize=2, eps=0.0, pvalueCutoff=1, pAdjustMethod="BH", gson=NULL, TERM2GENE = bckgr_geneset_Stress, TERM2NAME = pathways_BP_terms, seed=TRUE, by="fgsea")


```
