# Code Project Repository (CPR)
---
## Description preface

---
## Overview of tools used
- QC : MultiQC and fastqc
- Trimming : fastp
- Mapping : HiSAT2 using BW algorithm with graph based indexing(open source binary available in following [link with install instructions](http://daehwankimlab.github.io/hisat2/download/)) 
- Sorting bam : samtools
- Quantification of expression : HTSeq - HTSeqCount, NormalizationPipeline.sh (R source)
- Differential expression analysis : DeSeq2, edgeR
- Gene enrichment : clusterProfiler, bioMaRt

---
## General Information

| Document Summary||
|-----------------|--------|
| Project| Bnapus_ME TFM |
| Date | - |
|Summary | A collection of the methods to perform RNAseq analysis pipeline|

### List of Authors

| Author name| Email|
|-----------------:|-----------|
|Natalia García Sánchez | natalia.garcia.sanchez@alumnos.upm.es |


## Code summary

| Folder  | Code's aim |
| - |-|
| Annotation retrieval | Retrieval of annotations of miRNA in genomic gff, ATG autophagy related genes, Oxidative stress genes |
| QC and Filtering |  |
| Mapping |  |
| Quantification of expression | |
| DE analysis | |


## Data summary

|Version|Data source name | Detailed description | 
|-----------------|--------|----------------------|
|v.0.0.1|||
|v.0.0.1|| |
|v.0.0.1|||
|v.0.0.1|||
